### ISSUE TITLE
Make into a plugin for fastbench ?

### ISSUE THREAD

**Issue by davidmarkclements** [mcollina/bhdr#1]

**Comment by davidmarkclements** [comment]
Generally would be a shame to divide usage - or lose fastbench momentum

**Comment by jsumners** [comment]
@davidmarkclements I believe the plan is to figure out if this is viable with `bhdr` and, if so, probably move it into `fastbench`.

**Comment by mcollina** [comment]
there is a problem with `fastbench`, and it is that it currently does not keep tracks of the individual time for each ops. This is implemented here by leveraging native-hdr-histogram, which is a binary module.

I would like to keep fastbench as simple as possible, without the need to support binaries.

**Comment by jsumners** [comment]
So you have decided against folding this into `fastbench` then @mcollina?

**Comment by mcollina** [comment]
Not really. I would like to. It's very very complex, anyway. Ideas are welcomed!!!!

**Comment by jsumners** [comment]
I'm confused. If `bhdr` requires `native-hdr-histogram`, and you don't want `fastbench` to use binary modules, how could `bhdr` possibly be folded into `fastbench`?

I'm trying to decide if I should fork `runbench` into `fastbench-runner`. If `bhdr` will not be replacing `fastbench` then it will be worth it. Otherwise, not.

**Comment by mcollina** [comment]
@jsumners my goal is to have `native-hdr-histogram` as an `optionalDependency`, and maybe leverage https://www.npmjs.com/package/summary to compute the data in place of `native-hdr-histogram` if it is not available, or if exact data is required.

Does this make more sense?

**Comment by davidmarkclements** [comment]
the question is, I think, are fastbench and bhdr close enough in function that we're essentially competing with ourselves. If so, they should be merged. If they are unique enough then keep apart. 

For me, it seems they're very closely related - if that's so, we just need to figure out the technical part (binaries). Options would be

1. merge into fastbench, optional dep with fallbacks - as @mcollina as said
2. an optional addition to fastbench - which was the original suggestion (make into a plugin for fastbench).

**Comment by jsumners** [comment]
It makes more sense, at least. I thought the idea of using the histogram was to _very_ accurately capture the numbers; so accurate that vanilla JS can't possibly compete.

**Comment by mcollina** [comment]
@jsumners of vanilla JS is speed of those computation. hdr-histogram provide an _approximated_ value, up to the given numbers of decimals when creating the histogram. Providing *exact* value (up to the float precision) is _harder_ in Vanilla JS, and it will create a lot of overhead.

> an optional addition to fastbench - which was the original suggestion (make into a plugin for fastbench).

That is not possible, because fastbench measure only *total time* (at the moment), while this calculates individual time. Because of this, we can calculate the standard deviation, and perform some comparative analysis with statistical relevance (see #3).
That is not possible with fastbench in the current form, without a massive refactoring, that will lead to a similar solution to 1.

**Comment by davidmarkclements** [comment]
To me the most elegant solution is

```js
const fastbench = require('fastbench')({reporter: require('bhdr')})
```

Appreciate this could be a lot of work

**Comment by mcollina** [comment]
Here is what I think:

a) there is a cli that can handle newline-delimited-json data for benchmarks, more or less what is in #3, that can also compare two results
b) there is `fastbench`, that should be rewritten to support that protocol. This rewrite means getting rid of "whole time" measurements and move to "single op" measurements, as it works here.
c) there is `native-hdr-histogram`, that can be optionally used by `fastbench` if it is present to speed up benchmarks.

No need for options.

The most significant piece of work is making sure this module works **without** native-hdr-histogram, and then we can rebrand.

**Comment by mcollina** [comment]
Any of you want to help?

**Comment by jsumners** [comment]
I have a lot of work to do with https://github.com/pljson/pljson and https://github.com/jscas . I've also started writing a UPnP Internet Gateway Device server. Other than offering my opinion on things, the only thing I can really commit to on this project is updating `runbench` to support the new protocol and be its own module.

**Comment by davidmarkclements** [comment]
@mcollina lets do a call

**Comment by mcollina** [comment]
@jsumners would you mind having a look at the protocol used in #3? It's based on newline delimited JSON.

**Comment by jsumners** [comment]
@mcollina it looks acceptable to me. It would very easily allow presenting the information by grouping the results under a) the `benchmark` and b) the `name`.
